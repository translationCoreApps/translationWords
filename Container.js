import React from 'react';
import View from './View.js';
// Constant declarations
const NAMESPACE = "translationWords";

class Container extends React.Component {
  constructor() {
    super();
    this.state = {
      showHelps: true
    };
  }

  componentWillMount() {
    let { ScripturePane } = this.props.settingsReducer.toolsSettings;
    if (!ScripturePane) {
      // initializing the ScripturePane settings if not found.
      this.props.actions.setToolSettings("ScripturePane", "currentPaneSettings", ["ulb"]);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.contextIdReducer && this.props.contextIdReducer !== nextProps.contextIdReducer) {
      let articleId = nextProps.contextIdReducer.contextId.groupId;
      nextProps.actions.loadResourceArticle('translationWords', articleId);
    }
  }

  toggleHelps() {
    this.setState({ showHelps: !this.state.showHelps });
  }

  render() {
    let view = <div />;
    let { contextId } = this.props.contextIdReducer;
    if (contextId !== null) {
      view = <View
        {...this.props}
        showHelps={this.state.showHelps}
        toggleHelps={this.toggleHelps.bind(this)}
      />
    }
    return view;
  }
}

module.exports = {
  name: NAMESPACE,
  container: Container
};
