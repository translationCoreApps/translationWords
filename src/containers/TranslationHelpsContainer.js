import React from 'react';
import PropTypes from 'prop-types';
import {TranslationHelps} from 'tc-ui-toolkit';
// helpers
import * as tHelpsHelpers from '../helpers/tHelpsHelpers';

class TranslationHelpsContainer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      showHelpsModal: false,
      modalArticle: '',
      articleCategory: ''
    };

    this.toggleHelpsModal = this.toggleHelpsModal.bind(this);
    this.followTHelpsLink = this.followTHelpsLink.bind(this);
    window.followLink = this.followTHelpsLink;
  }

  componentWillMount() {
    this._reloadArticle(this.props);
  }

  componentWillReceiveProps(nextProps) {
    const {contextIdReducer, resourcesReducer} = this.props || {};
    const nextContextIDReducer = nextProps.contextIdReducer;
    if (contextIdReducer !== nextContextIDReducer) {
      this._reloadArticle(nextProps);
    }

    const {contextId} = contextIdReducer;
    const nextContextId = nextContextIDReducer.contextId;

    const currentArticle = tHelpsHelpers.getArticleFromState(resourcesReducer, contextId);
    const nextArticle = tHelpsHelpers.getArticleFromState(nextProps.resourcesReducer, nextContextId);
    if (currentArticle !== nextArticle) {
      var page = document.getElementById("helpsbody");
      if (page) page.scrollTop = 0;
    }
  }

  toggleHelpsModal() {
    this.setState({
      showHelpsModal: !this.state.showHelpsModal,
      modalArticle: ''
    });
  }

  /**
   * Loads the resource article
   * @param props
   * @private
   */
  _reloadArticle(props) {
    const {contextIdReducer, toolsReducer, currentProjectToolsSelectedGL, actions} = props;
    const {contextId} = contextIdReducer;
    if (contextId) {
      const articleId = contextId.groupId;
      const {currentToolName} = toolsReducer;
      const languageId = currentProjectToolsSelectedGL[currentToolName];
      actions.loadResourceArticle(currentToolName, articleId, languageId);
    }
  }

  followTHelpsLink(link) {
    let linkParts = link.split('/'); // link format: <lang>/<resource>/<category>/<article>

    const [lang, type, category, article] = linkParts;
    const resourceDir = tHelpsHelpers.getResourceDirByType(type);

    this.props.actions.loadResourceArticle(resourceDir, article, lang, category);
    const articleData = this.props.resourcesReducer.translationHelps[resourceDir][article];

    let newState;
    const showHelpsModal = true;
    const articleCategory = category;
    if (articleData) {
      newState = {
        showHelpsModal,
        articleCategory,
        modalArticle: articleData
      };
    } else {
      newState = {
        showHelpsModal,
        articleCategory,
        modalArticle: 'Cannot find an article for ' + link
      };
    }
    //todo: Shouldn't need to to set state and return state in the same function
    // Seems like an anti pattern
    this.setState(newState);
    return newState;
  }

  render() {
    const {
      currentProjectToolsSelectedGL,
      toolsReducer: {currentToolName},
      resourcesReducer,
      contextIdReducer: {contextId},
      showHelps,
      toggleHelps,
      translate
    } = this.props;
    const languageId = currentProjectToolsSelectedGL[currentToolName];
    const currentFile = tHelpsHelpers.getArticleFromState(resourcesReducer, contextId);
    const currentFileMarkdown = tHelpsHelpers.convertMarkdownLinks(currentFile, languageId);
    const tHelpsModalMarkdown = tHelpsHelpers.convertMarkdownLinks(this.state.modalArticle, languageId, this.state.articleCategory);
    return (
      <TranslationHelps
        translate={translate}
        article={currentFileMarkdown}
        modalArticle={tHelpsModalMarkdown}
        openExpandedHelpsModal={() => this.toggleHelpsModal()}
        isShowHelpsSidebar={showHelps}
        sidebarToggle={toggleHelps}
        isShowHelpsExpanded={this.state.showHelpsModal} />
    );
  }
}

TranslationHelpsContainer.propTypes = {
  currentProjectToolsSelectedGL: PropTypes.object,
  translate: PropTypes.func,
  resourcesReducer: PropTypes.object,
  contextIdReducer: PropTypes.shape({
    contextId: PropTypes.object.isRequired
  }),
  toolsReducer: PropTypes.object,
  actions: PropTypes.shape({
    loadResourceArticle: PropTypes.func.isRequired,
  }),
  showHelps: PropTypes.bool.isRequired,
  toggleHelps: PropTypes.func.isRequired,
};

export default TranslationHelpsContainer;
