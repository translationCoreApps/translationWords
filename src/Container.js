/* eslint-env jest */
import {connect} from 'react-redux';
import React from 'react';
import PropTypes from 'prop-types';
//selectors
import {
  getContextId,
  getManifest,
  getCurrentProjectToolsSelectedGL,
  getGroupsIndex,
  getResourceByName,
  getSelections,
  getCurrentPaneSettings,
  getBibles
} from './selectors';
// helpers
import * as settingsHelper from './helpers/settingsHelper';
//containers
import GroupMenuContainer from './containers/GroupMenuContainer';
import VerseCheckContainer from './containers/VerseCheckContainer';
import TranslationHelpsContainer from './containers/TranslationHelpsContainer';
import CheckInfoCardContainer from './containers/CheckInfoCardContainer';
import ScripturePaneContainer from './containers/ScripturePaneContainer';


class Container extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      showHelps: true
    };
    this.toggleHelps = this.toggleHelps.bind(this);
  }
  componentWillMount() {
    const { bibles } = this.props.scripturePane;
    settingsHelper.loadCorrectPaneSettings(this.props, this.props.tc.actions.setToolSettings, bibles);
  }

  toggleHelps() {
    this.setState({showHelps: !this.state.showHelps});
  }

  render() {
    const {
      contextIdReducer: {contextId}
    } = this.props;

    if (contextId !== null) {
      // const glQuote = actions.getGLQuote(languageId, groupId, currentToolName);
      return (
        <div style={{display: 'flex', flexDirection: 'row', width: '100vw'}}>
          <GroupMenuContainer {...this.props.groupMenu} />
          <div style={{display: 'flex', flexDirection: 'column', width: '100%', overflowX: 'auto'}}>
            <div style={{ height: '250px', paddingBottom: '20px' }}>
              <ScripturePaneContainer {...this.props.scripturePane} />
            </div>
            <CheckInfoCardContainer
              toggleHelps={this.toggleHelps.bind(this)}
              showHelps={this.state.showHelps}
              {...this.props.checkInfoCard} />
            <VerseCheckContainer {...this.props.verseCheck} />
          </div>
          <TranslationHelpsContainer
            toggleHelps={this.toggleHelps.bind(this)}
            showHelps={this.state.showHelps}
            {...this.props.translationHelps} />
        </div>
      );
    } else {
      return null;
    }
  }
}

Container.propTypes = {
  translationHelps: PropTypes.any,
  groupMenu: PropTypes.any,
  verseCheck: PropTypes.any,
  checkInfoCard: PropTypes.any,
  translate: PropTypes.func,
  settingsReducer: PropTypes.shape({
    toolsSettings: PropTypes.shape({
      ScripturePane: PropTypes.object
    })
  }),
  contextIdReducer: PropTypes.shape({
    contextId: PropTypes.shape({
      groupId: PropTypes.any
    })
  }),
  groupsIndexReducer: PropTypes.shape({
    groupsIndex: PropTypes.array
  }),
  projectDetailsReducer: PropTypes.shape({
    currentProjectToolsSelectedGL: PropTypes.object.isRequired
  }),
  tc: PropTypes.shape({
    actions: PropTypes.shape({
      setToolSettings: PropTypes.func.isRequired,
      loadResourceArticle: PropTypes.func.isRequired,
      getGLQuote: PropTypes.func.isRequired,
      getSelectionsFromContextId: PropTypes.func.isRequired
    })
  }),
  scripturePane: PropTypes.object.isRequired
};

const mapStateToProps = (state, ownProps) => {
  const legacyToolsReducer = {currentToolName: ownProps.tc.selectedToolName};
  return {
    groupMenu: {
      tc: ownProps.tc,
      groupsDataReducer: ownProps.tc.groupsDataReducer,
      groupsIndexReducer: ownProps.tc.groupsIndexReducer,
      translate: ownProps.translate
    },
    verseCheck: {
      translate: ownProps.translate,
      currentToolName: ownProps.tc.selectedToolName,
      projectDetailsReducer: ownProps.tc.projectDetailsReducer,
      loginReducer: ownProps.tc.loginReducer,
      resourcesReducer: ownProps.tc.resourcesReducer,
      commentsReducer: ownProps.tc.commentsReducer,
      selectionsReducer: ownProps.tc.selectionsReducer,
      contextIdReducer: ownProps.tc.contextIdReducer,
      toolsReducer: legacyToolsReducer,
      groupsDataReducer: ownProps.tc.groupsDataReducer,
      remindersReducer: ownProps.tc.remindersReducer,
      actions: ownProps.tc.actions
    },
    translationHelps: {
      translate: ownProps.translate,
      currentProjectToolsSelectedGL: getCurrentProjectToolsSelectedGL(ownProps),
      toolsReducer: legacyToolsReducer,
      resourcesReducer: ownProps.tc.resourcesReducer,
      contextIdReducer: ownProps.tc.contextIdReducer,
      actions: ownProps.tc.actions
    },
    checkInfoCard: {
      translate: ownProps.translate,
      translationHelps: getResourceByName(ownProps, 'translationHelps'),
      groupsIndex: getGroupsIndex(ownProps),
      contextId: getContextId(ownProps)
    },
    scripturePane: {
      translate: ownProps.translate,
      manifest: getManifest(ownProps),
      selections: getSelections(ownProps),
      currentPaneSettings: getCurrentPaneSettings(ownProps),
      bibles: getBibles(ownProps),
      contextId: getContextId(ownProps),
      projectDetailsReducer: ownProps.tc.projectDetailsReducer,
      showPopover: ownProps.tc.actions.showPopover,
      editTargetVerse: ownProps.tc.actions.editTargetVerse,
      getLexiconData: ownProps.tc.actions.getLexiconData,
      setToolSettings: ownProps.tc.actions.setToolSettings,
      getAvailableScripturePaneSelections: ownProps.tc.actions.getAvailableScripturePaneSelections,
      makeSureBiblesLoadedForTool: ownProps.tc.actions.makeSureBiblesLoadedForTool
    }
  };
};


export default connect(mapStateToProps)(Container);
