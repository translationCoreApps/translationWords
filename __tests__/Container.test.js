/* eslint-env jest */
import React from 'react';
import Container from '../src/Container';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import renderer from 'react-test-renderer';
import {shallow, mount, render} from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import {Provider} from 'react-redux';
import toJson from 'enzyme-to-json';
import TranslationHelpsContainer from '../src/containers/TranslationHelpsContainer';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const basicProps = require('./__fixtures__/basicProps.json');
const props = {
  ...basicProps,
  translate: k => k,
  tc: {
    ...basicProps.tc,
    project: {
      getBookName: () => () => "gen"
    }
  }
};

props.tc.actions = {
  showPopover: jest.fn(),
  editTargetVerse: jest.fn(),
  getLexiconData: jest.fn(),
  setToolSettings: jest.fn(),
  loadResourceArticle: jest.fn(),
  getGLQuote: jest.fn(),
  setFilter: jest.fn(),
  changeSelections: jest.fn(),
  goToNext: jest.fn(),
  goToPrevious: jest.fn(),
  getAvailableScripturePaneSelections: jest.fn(),
  makeSureBiblesLoadedForTool: jest.fn(),
  groupMenuChangeGroup: jest.fn(),
  groupMenuExpandSubMenu: jest.fn(),
  getSelectionsFromContextId: () => ''
};

const store = mockStore({});
describe.only('Container Tests', () => {
  it('Test Container', () => {
    const wrapper = render(
      <MuiThemeProvider>
        <Provider store={store}>
          <Container {...props} />
        </Provider>
      </MuiThemeProvider>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Test empty Container', () => {
    const myProps = {
      ...props,
      contextIdReducer: {
        contextId: null
      }
    };
    const wrapper = renderer.create(
      <MuiThemeProvider>
        <Provider store={store}>
          <Container {...myProps} />
        </Provider>
      </MuiThemeProvider>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Test Container.toggleHelps', () => {
    const container = shallow(
      <Provider store={store}>
        <Container {...props} />
      </Provider>)
      //wierd stuff to get the inner component
      .dive({context: {store}}).dive().instance();
    expect(container.state.showHelps).toBeTruthy();
    container.toggleHelps();
    expect(container.state.showHelps).toBeFalsy();
    container.toggleHelps();
    expect(container.state.showHelps).toBeTruthy();
  });

  it('Test has ScripturePane Panes', () => {
    const container = render(
      <Provider store={store}>
        <Container {...props} />
      </Provider>)
      //wierd stuff to get the inner component
      .find('.panes-container').children().length;
    expect(container).toBe(4);
  });

  it('Test does not have ScripturePane', () => {
    props.tc.bibles = {};
    const container = shallow(
      <Provider store={store}>
        <Container {...props} />
      </Provider>)
      //wierd stuff to get the inner component
      .find('.panes-container').children().length;
    expect(container).toBe(0);
  });

  it('Test TranslationHelps componentWillReceiveProps', () => {
    const root = render(
      <Provider store={store}>
        <Container {...props} />
      </Provider>)
      .find(TranslationHelpsContainer);
      expect(props.tc.actions.loadResourceArticle).toHaveBeenCalledWith("translationWords", "blasphemy", "en");
  });
});
