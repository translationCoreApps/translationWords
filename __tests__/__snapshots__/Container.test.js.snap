// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container Tests Test Container 1`] = `
<[View]
  actions={
    Object {
      "loadResourceArticle": [Function],
      "setToolSettings": [Function],
    }
  }
  contextIdReducer={
    Object {
      "contextId": Object {
        "groupId": "groupId1",
      },
    }
  }
  groupsIndexReducer={
    Object {
      "groupsIndex": Array [
        Object {
          "id": "groupId1",
          "name": "group title",
        },
        Object {
          "id": "groupId2",
          "name": "group title2",
        },
      ],
    }
  }
  projectDetailsReducer={
    Object {
      "currentProjectToolsSelectedGL": Object {
        "translationWords": "en",
      }
    }
  }
  settingsReducer={
    Object {
      "toolsSettings": Object {
        "ScripturePane": Object {
          "currentPaneSettings": Array [
            Object {
              "bibleId": "ugnt",
              "languageId": "originalLanguage",
            },
          ],
        },
      },
    }
  }
  showHelps={true}
  title="group title"
  toggleHelps={[Function]}
  toolsReducer={
    Object {
      "currentToolName": "translationWords",
    }
  }
  translate={[Function]}
/>
`;

exports[`Container Tests Test empty Container 1`] = `null`;
